### ISSUE TITLE
4-bytes utf8 char validation

### ISSUE THREAD

**Issue by boost2017** [Gumichan01/utf8_string#1]
The way you validate 4-bytes utf8 character is questionable.
The first of 4-bytes char must be between 0xf0 and 0xf4, which you forget to test. 
```
            // If the first byte of the sequence is 0xF0
            // then the first continuation byte must be between 0x90 and 0xBF
            // otherwise, if the byte is 0xF4
            // then the first continuation byte must be between 0x80 and 0x8F
            if(*it == '\xF0')
            {
                if(*(it + 1) < '\x90' || *(it + 1) > '\xBF')
                    return false;
            }
            else if(*it == '\xF4')
            {
                if(*(it + 1) > '\x8F')
                    return false;
            }
```

**Comment by Gumichan01** [comment]
Hey,

Sorry for taking a so long time to answer (^_^').

This piece of code is just dealing with particular cases (the first byte is 0xF0 or 0xF4). At this point, I have already checked that my first byte is between 0xF0 and 0xF4.

Let me show you the entire code related to the 4-bytes utf-8 character
```cpp
        // ...
        if((0xF8&  *it) == 0xF0)
        {
            // The UTF-8 codepoint begin with 0b11110xxx -> 4-byte codepoint
            // If the iterator reach the end of the string before the
            // end of the 4-byte codepoint -> invalid string
            if((it + 1) == itend || (it + 2) == itend || (it + 3) == itend)
                return false;

            // Each of the following bytes is a value
            // between 0x80 and 0xBF
            if(((0xC0&  *(it + 1)) != 0x80) || ((0xC0&  *(it + 2)) != 0x80)
                    || ((0xC0&  *(it + 3)) != 0x80))
            {
                return false;
            }

            // If the first byte of the sequence is 0xF0
            // then the first continuation byte must be between 0x90 and 0xBF
            // otherwise, if the byte is 0xF4
            // then the first continuation byte must be between 0x80 and 0x8F
            if(*it == '\xF0')
            {
                if(*(it + 1) < '\x90' || *(it + 1) > '\xBF')
                    return false;
            }
            else if(*it == '\xF4')
            {
                if(*(it + 1) > '\x8F')
                    return false;
            }

            it += 4; // Jump to the next codepoint
        }
        // ...
```

This line of code 
```cpp
if ( (0xF8 & *it) == 0xF0 )
```
tests if the first 4-bytes char is between 0xF0 and 0xF4. Actually, this comparison is not correct. Indeed, if the first 4-bytes character is one of the following: ```` 0xF5, 0xF6, 0xF7 ```` the condition becomes *true*, but it shouldn't.

This is a bug to fix.

**Comment by Gumichan01** [comment]
This bug has been fixed in 141df48
